File: day24_part2_2021.py

import os

def read_all(path):
    with open(path, 'r') as file:
        return file.read().strip()

def num(w):
    n = 0
    for i in range(len(w)):
        n *= 10
        n += w[i]
    return n

def main():
    k, l, m = [], [], []
    for i, line in enumerate(read_all("input.txt").split("\n")):
        v = 0
        if i % 18 == 4:
            v = int(line.split()[2])
            l.append(v)
        elif i % 18 == 5:
            v = int(line.split()[2])
            k.append(v)
        elif i % 18 == 15:
            v = int(line.split()[2])
            m.append(v)

    constraints = {}
    stack = []
    for i in range(len(l)):
        if l[i] == 1:
            stack.append(i)
        elif l[i] == 26:
            pop = stack.pop()
            constraints[pop] = [i, m[pop] + k[i]]

    min_val = [0] * 14
    for i in range(14):
        if i not in constraints:
            continue
        vmin = 1
        while vmin + constraints[i][1] < 1:
            vmin += 1
        min_val[i] = vmin
        min_val[constraints[i][0]] = vmin + constraints[i][1]

    print(num(min_val))

if __name__ == "__main__":
    main()

File: java/day24_part2_2021.java

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Stack;

public class Solution {

  public static String readAll(String path) throws IOException {
    return Files.readString(Paths.get(path)).strip();
  }

  public static long num(int[] w) {
    long n = 0;
    for (int digit : w) {
      n = n * 10 + digit;
    }
    return n;
  }

  public static void main(String[] args) throws IOException {
    List<Integer> k = new ArrayList<>();
    List<Integer> l = new ArrayList<>();
    List<Integer> m = new ArrayList<>();
    String[] lines = readAll("input.txt").split("\n");
    for (int i = 0; i < lines.length; i++) {
      int v;
      if (i % 18 == 4) {
        v = Integer.parseInt(lines[i].split(" ")[2]);
        l.add(v);
      } else if (i % 18 == 5) {
        v = Integer.parseInt(lines[i].split(" ")[2]);
        k.add(v);
      } else if (i % 18 == 15) {
        v = Integer.parseInt(lines[i].split(" ")[2]);
        m.add(v);
      }
    }

    Map<Integer, int[]> constraints = new HashMap<>();
    Stack<Integer> stack = new Stack<>();
    for (int i = 0; i < l.size(); i++) {
      if (l.get(i) == 1) {
        stack.push(i);
      } else if (l.get(i) == 26) {
        int pop = stack.pop();
        constraints.put(pop, new int[] {i, m.get(pop) + k.get(i)});
      }
    }

    int[] minVal = new int[14];
    for (int i = 0; i < 14; i++) {
      if (!constraints.containsKey(i)) {
        continue;
      }
      int vmin = 1;
      while (vmin + constraints.get(i)[1] < 1) {
        vmin++;
      }
      minVal[i] = vmin;
      minVal[constraints.get(i)[0]] = vmin + constraints.get(i)[1];
    }
    System.out.println(num(minVal));
  }
}
